### PR TITLE
FIX: server subscriptions affecting each other

### LINKED ISSUE #813: Array slice channel filter works intermittently when used with camonitor
## Issue

Using the [array slice channel filter](https://epics.anl.gov/base/R3-15/1-docs/filters.html#Array-Filter-arr) shortcut to grab the first 5 elements of a large waveform:

```
$ camonitor IOC:TST:DScan:Qmini:SPECTRUM.[:5] IOC:TST:DScan:Qmini:WAVELENGTHS.[:5]
IOC:TST:DScan:Qmini:SPECTRUM.[:5] 2023-05-17 14:20:38.572136 6 0.000278906 0.387099 0.0385469 0.0377823 0.0375703 0.0375969
IOC:TST:DScan:Qmini:WAVELENGTHS.[:5] 2023-05-17 14:20:24.507546 6 0.186867 0.187339 0.187811 0.188283 0.188755 0.189227
```

This occasionally fails, yielding the entire waveform:

```
$ camonitor IOC:TST:DScan:Qmini:SPECTRUM.[:5] IOC:TST:DScan:Qmini:WAVELENGTHS.[:5]
IOC:TST:DScan:Qmini:SPECTRUM.[:5] 2023-05-17 14:20:56.844536 6 0.000319747 0.387147 0.0385906 0.0378262 0.0376146 0.0376449
IOC:TST:DScan:Qmini:WAVELENGTHS.[:5] 2023-05-17 14:20:24.507546 6 0.186867 0.187339 0.187811 0.188283 0.188755 0.189227
IOC:TST:DScan:Qmini:SPECTRUM.[:5] 2023-05-17 14:20:57.045808 2048 0.000320493 0.387147 0.0385915 0.0378265 0.0376147 0.0376451 0.0374921 0.0374403 0.0375272 0.0378044 0.0378637 0.0380535 0.0382612 0.0384328 0.0388858 0.0387976 0.0392194 0.0394759 0.0396871 0.0400781 0.0402357 0.0404092 0.0407816 0.0413916 0.0425398 0.0426763 0.0428687 0.0430999 0.0434778 0.0434827 0.0437234 0.0439246 0.0440915 0.0442264 0.0442996 0.0444089 0.0445642 (... clipped)
```

This is likely also the case with the other supported channel filters, but I have only tested (or really cared to use) this one.

Reproducible with `caproto-monitor` as well - should be better to help debug the issue.

## Code reference

Channel filter parsing:

https://github.com/caproto/caproto/blob/20d59991651b1628db4f6f44115a9841dfc90e2c/caproto/_utils.py#L710

Where the filter should be applied to the data prior to sending it to the circuit:

https://github.com/caproto/caproto/blob/20d59991651b1628db4f6f44115a9841dfc90e2c/caproto/server/common.py#L922

## Environment

caproto v1.0.1

### PR DESCRIPTION
**Pull request by klauer**
Closes #813 

See the linked issue for further details.

Tested and found to work locally:
* The array slice filter (with two separate monitors)
* The deadband filter

This is fundamental code to caproto's server mechanism, so a detailed/thorough review would be wise and appreciated.

### CODE REVIEW

**Comment by klauer** [comment]
The original commit inadvertently broke API that I didn't realize trio was relying on. I, um, un-cleaned it a bit and re-added the old method and now important tests are passing.

Semi-related, as it's due to the subscription mechanism: we need to look into https://github.com/caproto/caproto/issues/797 at some point. (Unless this somehow magically fixes it...)

File: caproto/server/common.py [hunk 1]
```diff
@@ -690,6 +690,8 @@ async def handle_write():
 
 
 class Context:
+    subscriptions: DefaultDict[SubscriptionSpec, Deque[Subscription]]
```
Comment (danielballan) [inline caproto/server/common.py:693]: Is it best practice to define a class var for typing purposes and then shadow it with an instance var in `__Init__`? My impulse would have been to define this with `self.subscriptions` but I am very early on the learning curve with Python typing.

**Review by danielballan — COMMENTED** [review]
I agree the refactor makes this more readable, and it's clear how we should have been passing `sub_spec` down into the filter down, not taking the last `sub_spec` in the loop.

One small ignorant question about types.

**Review by danielballan — COMMENTED** [review]
I agree the refactor makes this more readable, and it's clear how we should have been passing `sub_spec` down into the filter down, not taking the last `sub_spec` in the loop.

One small ignorant question about types.

**Review by danielballan — COMMENTED** [review]
I agree the refactor makes this more readable, and it's clear how we should have been passing `sub_spec` down into the filter down, not taking the last `sub_spec` in the loop.

One small ignorant question about types.

File: caproto/server/common.py (hunk 1 above)
Comment (klauer) [reply caproto/server/common.py:693]: This just looks out of place, being the only thing annotated...
It's not actually defining a classvar - or any variable for that matter.  As written, it's just a type hint that says "`Context` instances have `subscriptions` which are defaultdicts".

Defining a classvar would look like the following:
```python
    subscriptions: ClassVar[DefaultDict[SubscriptionSpec, Deque[Subscription]]] = defaultdict(...)
```

**Comment by klauer** [comment]
(Merging in Azure CI fix)

**Comment by klauer** [comment]
@tacaswell @clintonroy have time for a review? Would like to merge and ready for a tag

**Review by tacaswell — APPROVED** [review]

**Comment by klauer** [comment]
@tacaswell rubber stamp or actual review?

**Comment by klauer** [comment]
Hard to find good help these days, or those that respond to simple questions even. *sigh*

I'll merge and take the blame as per usual

**Comment by tacaswell** [comment]
it was an actual review.

**Comment by tacaswell** [comment]
Also, it was dinner time on the east coast!

**Comment by klauer** [comment]
Sorry for being a doubting Ken, Thomas! I know you understand the feeling of screaming into the wind when it comes to open source projects. 

caproto is one of the most frustrating ones in that regard for me, and it shows a bit too much in my replies here.

**Pull request merged**